app/controllers/supplementary/TransportInformationContainersPageController.scala

```diff
@@ -0,0 +1,131 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.supplementary
+
+import config.AppConfig
+import controllers.actions.AuthAction
+import controllers.supplementary.routes.{TotalNumberOfItemsController, TransportInformationContainersPageController}
+import controllers.util.CacheIdGenerator.supplementaryCacheId
+import controllers.util.MultipleItemsHelper.{add, remove, saveAndContinue}
+import controllers.util.{Add, FormAction, Remove, SaveAndContinue}
+import forms.supplementary.TransportInformationContainer
+import forms.supplementary.TransportInformationContainer.form
+import handlers.ErrorHandler
+import javax.inject.Inject
+import models.declaration.supplementary.TransportInformationContainerData
+import models.declaration.supplementary.TransportInformationContainerData.{id, maxNumberOfItems}
+import models.requests.AuthenticatedRequest
+import play.api.data.Form
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent}
+import services.CustomsCacheService
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import views.html.supplementary.add_transport_containers
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class TransportInformationContainersPageController @Inject()(
+  override val messagesApi: MessagesApi,
+  authenticate: AuthAction,
+  errorHandler: ErrorHandler,
+  customsCacheService: CustomsCacheService
+)(implicit ec: ExecutionContext, appConfig: AppConfig)
+    extends FrontendController with I18nSupport {
+
+  implicit val countries = services.Countries.allCountries
+
+  def displayPage(): Action[AnyContent] = authenticate.async { implicit request =>
+    customsCacheService
+      .fetchAndGetEntry[TransportInformationContainerData](supplementaryCacheId, id)
+      .map {
+        case Some(data) => Ok(add_transport_containers(form, data.containers))
+        case _          => Ok(add_transport_containers(form, Seq()))
+      }
+  }
+
+  def handlePost(): Action[AnyContent] = authenticate.async { implicit request =>
+    val boundForm = form.bindFromRequest()
+
+    val actionTypeOpt = request.body.asFormUrlEncoded.map(FormAction.fromUrlEncoded(_))
+
+    val cachedData = customsCacheService
+      .fetchAndGetEntry[TransportInformationContainerData](supplementaryCacheId, id)
+      .map(_.getOrElse(TransportInformationContainerData(Seq())))
+
+    cachedData.flatMap { cache =>
+      actionTypeOpt match {
+        case Some(Add) => addContainer(boundForm, maxNumberOfItems, cache)
```

Comment:
you can make the cache Implicit

Reply:
I actually like having the `cachedData` as an explicity variable as this is business logic. 
I like the use of `implicit` for `currying` or to pass boiler plate parameters (Play's `request`, HMRC's middleware, app config, etc), something that does not immediately translates as business logic, but more as app/infrastructure dependencies, something that a developer should not need to worry as such, as they are always there regardless of the logic needs/requirements.

Reply:
@raghu1978 What do you think?

Reply:
in this case it is implicit that you need the collection and to pass it across you still are aware in the code as every method has that as an implicit parmater, there is no difference in how we pass header carrier and execution context

Reply:
For example https://github.com/hmrc/customs-declare-exports-frontend/blob/master/test/generators/Generators.scala#L107 this is something I dislike.
1. It is something that is only going to be used by the one `form`  but it is instead in a generic Generators class.
2. It is `implicit` when there is not need to such and makes life of the tester/developer a bit harder when could have well been defined in the test method or as a function on the companion spec object.

Reply:
>  there is no difference in how we pass header carrier and execution context
None at all, the difference is not how we pass it, but how we use them. The `request`, `appConfig`, `headerCarrier`, `format`, `executionContext` it makes some sense to pass them as implicit variables (the same with Authorization/Middleware libraries/functions or as I said before `currying`), although for example I would think twice about passing `countries` as an implicit variable.

The somewhat obsession of writing less code, does not translate into more practical or readable code, it just becomes more _enigmatic_. The `cachedData` is a variable that holds a state that is  useuful to _Transport Containers_ business logic, so I would like to continue to explicitly using it.

Another good justification to use implicit variables could be because the signature of the function is now too verbose or too long which is also not the case.

In EffectiveScala they seems to corroborate with me http://twitter.github.io/effectivescala/#Types%20and%20Generics-Implicits and I have also brought up this discussion in the #community-scala and people do not strongly disagree with it.

I hope that this `implicit` use does not block the PR.

Reply:
Cache shouldn't be passed as implicit. I told you this in your last PR or during talking with you. Countries is something that always is constant. In my opinion it can be passed as implicit. Cache shouldn't. It's my opinion.